### PR TITLE
F2 b for paper trading

### PR DESCRIPTION
**Pull request by XavierOlsen**
Merge pull request https://github.com/IUS-CS/project-midwest-market-machines/pull/80 from IUS-CS/database-handler
Issues from previous branch are now addressed:
## Database Handler
* #pragma once instead of ifndef
* change holdings database name to holdings.csv
* enable rudimentary holdings operations, adding, removing, modifying user coins, largely untested
* Uses JSON pointer instead of stringstream for passing transactions from frontend. 
* Stringstream logic was moved to Database_Handler.h from the Binance_Websockets onMessageCallback
* Added some aliases to ease cogntive load
* Verified file does not infinite loop, terminates on EOF
* Added separate transactionHistory csv file to store
## Binance_Websockets Frontend to Backend
* Resolved thread collision issue with Binance and Database Traffic
* Trimmed unnecessary safeguards and try-catchs; program is locally hosted
* Adopted aliases, trimmed and clarified comments
* Ported Stringstream functionality to Database Handler.

## New functionality to look at has been introduced:
- Another Websocket Channel on 8081 
- Placeholder transactionHistory and holdings .csvs. Will remove and create tests to verify bounds checking code that prevents corruption crashing program works
- External tagging on JSONs coming from Database_Handler.h allows for logic based on JSON's origin
- Historical Data sends to frontend through 8081
- Will verify relative path correctly finds and writes to file
- useCryptoSocket and App.jsx placeholder buttons and websocket behavior
image

### CODE REVIEW

File: src/Binance_Websockets.cpp [hunk 1]
```diff
@@ -41,23 +45,52 @@ WebSocketServer Server(8080, "127.0.0.1");
...
-         const MessagePtr &msg) {});
-
-  // Blocking call - need to explicitly call Server.stop() later.
-  Server.listenAndStart();
+Server.setOnClientMessageCallback(
+[](shared_ptr<ix::ConnectionState> connectionState, WebSocket &webSocket, const MessagePtr &msg) {
+    if (msg->type == ix::WebSocketMessageType::Open) {
+        Database_Handler db;
+        db.sendHoldingsData(webSocket);
+    }
```
Comment (hwpoole) [inline src/Binance_Websockets.cpp]: We don't start the front end until after we open a connection to Binance's WebSocket API. It is possible we may not be finished sending the user's current holdings by the time we start receiving, processing, and sending klines from Binance to the front end.

Check `int main() {...}` for the order-of-operations on when things are spawned and dispatched.

Do we have a way to prevent an overrun on the WebSocket to the front end? What happens if a client connects to Binance and starts receiving klines almost immediately? Surely they will send that kline while we are sending historical data.

Can we hold all received klines until the historical data is done? Or use some sort of flag between `Database_Handler.h` and `Binance_Websockets.cpp` to pause execution until the database is fully sent?

Maybe something like this:

```C++
// Spawn server
// Spawn WebView front end.

while (!db.isDone()) { // A boolean
    pthread_cond_wait(&DatabaseCondition, &DatabaseLocker);
}

// Spawn clients
```

Or use `nanosleep()` for 0.5-ish seconds in a while-not-done loop for a hack-y-er solution.

Comment (hwpoole) [reply src/Binance_Websockets.cpp]: Or, are we near-guaranteed to complete sending the user's holdings by the time Binance returns klines?

File: src/Binance_Websockets.cpp [hunk 2]
```diff
@@ -41,23 +45,52 @@ WebSocketServer Server(8080, "127.0.0.1");
...
+    if (msg->type == ix::WebSocketMessageType::Message) 
+    {
+        try {
+            nlohmann::json incoming = nlohmann::json::parse(msg->str);
+            if (incoming.contains("type")) {
+                string transactionType = incoming["type"];
```
Comment (hwpoole) [inline src/Binance_Websockets.cpp]: Need not assume possibility of a malformed JSON object.

File: src/Binance_Websockets.cpp [hunk 3]
```diff
@@ -41,23 +45,52 @@ WebSocketServer Server(8080, "127.0.0.1");
...
+                    transactionType << "," << 
+                    incoming["coin"] << "," << 
+                    incoming["price"] << "," << 
+                    incoming["quantity"] << "," << 
+                    time(0);
+                    db.recordTransaction(ss.str());
+                }
+            }
+        } catch (const std::exception &e) {
+            std::cerr << "Failed to parse message: " << msg->str << std::endl;}
```
Comment (hwpoole) [inline src/Binance_Websockets.cpp]: `try{...}catch` is unnecessary. In what case should this fail? See lines `150-160`.

File: src/Binance_Websockets.cpp [hunk 4]
```diff
@@ -41,23 +45,52 @@ WebSocketServer Server(8080, "127.0.0.1");
...
+    }
+});
+    Server.listenAndStart();
```
Comment (hwpoole) [inline src/Binance_Websockets.cpp]: Check out lines `31-37`. I included multiple aliases that should make your life easier here.

```C++
//  Quality of life statements. All are unneccessary, strictly speaking.
using json = nlohmann::json;
using MessageType = ix::WebSocketMessageType;
using WebSocket = ix::WebSocket;
using WebSocketServer = ix::WebSocketServer;
using MessagePtr = ix::WebSocketMessagePtr;
using WebView = webview::webview;
using namespace std;
```

As in:

1. You need not type `ix::WebSocketMessageType::Open`, you may instead type `MessageType::Open`
2. Same for `ix::WebSocketMessageType::Message`, vs. `MessageType::Message`.
3. You need not type `nlohmann::json incoming = nlohmann::json::parse(msg->str);`, you may instead use `json incoming = json::parse(msg->str);`

You correctly applied this in the arguments for `Server.setOnClientMessageCallback(...)`, where you used `WebSocket &webSocket` instead of `ix::WebSocket &webSocket`, and `const MessagePtr &msg` instead of `const ix:WebSocketMessagePtr &webSocket`.

These aliases work as shortcuts. There is no harm to ignore this comment and include these snippets as-is other than for styling. It's out-of-line with the rest of the file.

File: src/Binance_Websockets.cpp [hunk 5]
```diff
@@ -41,23 +45,52 @@ WebSocketServer Server(8080, "127.0.0.1");
...
+ * 
+ *Callback from the frontend of MessageType::Message writes valid transactions to the database.csv file 
+ * by contstruction a stringstream of the transaction data, and passing it to the Database_Handler's writeTransaction() method.
+ *  
+ * Valid transactions are JSON objects with a "type" field of 
+ * "buy" or "sell", 
+ * "coin", 
+ * "price",
+ * "quantity", and
+ * "time" - time(0) from time.h sends a simple UNIX timestamp, preventing the need for a separate time library like <chrono>.
```
Comment (hwpoole) [inline src/Binance_Websockets.cpp]: On first read of this comment block, I'm lead to believe these lines are part of the "type" field.

"Valid transactions are JSON objects with a "type" field of ..."

**Review by hwpoole — CHANGES_REQUESTED** [review]
Hello @XavierOlsen, and thanks for the new feature implementation.

I have some concerns about the order-of-operations on the database sending. Let me know what you think.

Thanks!

File: src/Binance_Websockets.cpp [hunk 6]
```diff
@@ -41,23 +45,52 @@ WebSocketServer Server(8080, "127.0.0.1");
...
+                string transactionType = incoming["type"];
+                if (transactionType == "buy" || transactionType == "sell") {
+                    Database_Handler db; std::stringstream ss;
+                    ss << 
+                    transactionType << "," << 
+                    incoming["coin"] << "," << 
+                    incoming["price"] << "," << 
+                    incoming["quantity"] << "," << 
+                    time(0);
+                    db.recordTransaction(ss.str());
```
Comment (hwpoole) [inline src/Binance_Websockets.cpp]: I made this comment on #74. The `Server.setOnClientMessageCallback(...) {...}` method should not be responsible for forming a StringStream. The `recordTransaction()` method should accept a JSON object and parse it into the required format. In the case that `recordTransaction()` accepts a StringStream and a StringStream only, it is now your responsibility to create a properly formatted StringStream in _every_ location that `recordTransaction()` is called.

In either case, you only need _one_ `Database_Handler db;` in this entire method. You can declare it at the top as follows:


```C++
Server.setOnClientMessageCallback(
[](shared_ptr<ix::ConnectionState> connectionState, WebSocket &webSocket, const MessagePtr &msg) {
    Database_Handler db;

    if (msg->type == MessageType::Open) {
        db.sendHoldingsData(webSocket);
    }
    // etc...
}
```

By following both recommendations, the `if(){...}` statement becomes:

```C++
if (transactionType == "buy" || transactionType == "sell") {
    db.recordTransaction(incoming);
}
```

File: src/Binance_Websockets.cpp (hunk 1 above)
Comment (XavierOlsen) [reply src/Binance_Websockets.cpp]: In your shadcn_ branch, I assumed this part was meant to stop that from happening,
image
I see what you're saying though, those messages would be sent to frontend and might end up in the chart data, the boolean seems like a good fix, I'll try it out

File: src/Binance_Websockets.cpp (hunk 4 above)
Comment (XavierOlsen) [reply src/Binance_Websockets.cpp]: Forgot that's what "using" was for, will clean it up

File: src/Binance_Websockets.cpp (hunk 6 above)
Comment (XavierOlsen) [reply src/Binance_Websockets.cpp]: True, that's a lot cleaner, now that we've got time this sprint I'll get it up to par to actually be merged into main instead of just minimal functionality

File: src/Binance_Websockets.cpp (hunk 1 above)
Comment (hwpoole) [reply src/Binance_Websockets.cpp]: Yes, that section of code in [useCryptoSocket.jsx](https://github.com/IUS-CS/project-midwest-market-machines/blob/shadcn_/MarketUI/src/useCryptoSocket.jsx) handles the flow like this:

1. Startup: assume all JSON objects sent are for the user's holdings.
2. Wait for the flag `last: true` to be sent in the last JSON object of the user's holdings.
3. Switch to expecting klines

File: src/Binance_Websockets.cpp (hunk 6 above)
Comment (hwpoole) [reply src/Binance_Websockets.cpp]: Pass a pointer to the object ideally, it'll be faster.

```C++
// Header file
inline void recordTransaction(json* transactionData) {
    // Parse it.
    // Remember to use arrows for pointers.
    // Arrows replace the '.' operator.
    Pointer->Method(stuff);
    New_Variable = Pointer->Property_of_pointer;
    
    // Get the time here too.
    Time = time(0);
    // Write it.
}

// Implementation
json *incoming = json::parse(msg->str);
```

File: src/Binance_Websockets.cpp (hunk 1 above)
Comment (hwpoole) [reply src/Binance_Websockets.cpp]: > Yes, that section of code in [useCryptoSocket.jsx](https://github.com/IUS-CS/project-midwest-market-machines/blob/shadcn_/MarketUI/src/useCryptoSocket.jsx) handles the flow like this:
> 
>     1. Startup: assume all JSON objects sent are for the user's holdings.
> 
>     2. Wait for the flag `last: true` to be sent in the last JSON object of the user's holdings.
> 
>     3. Switch to expecting klines

Notably, we do not check for what a JSON object's purpose is. We just make an assumption that we will receive all the holdings, and then we will receive some klines.

That could be changed.

Comment (XavierOlsen) [reply src/Binance_Websockets.cpp]: Should be resolved with new port

File: src/Binance_Websockets.cpp (hunk 6 above)
Comment (XavierOlsen) [reply src/Binance_Websockets.cpp]: Removed unnecessary double declaration of db, now passing a json pointer to database_handler

File: src/Database_Handler.h [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+using json = nlohmann::json;
+using namespace std;
+
+/* class Database_Handler
+ * Receives frontend transaction information through Binance_Websockets.cpp,
+ * Writes transactions to transactionHistory.csv and updates holdings.csv,
+ * On connection open, sends current holdings, transaction history, and
+ * historical price data to the frontend, with data from each wrapped with tags to allow for 
+ * differentiation on the frontend
```
Comment (hwpoole) [inline src/Database_Handler.h:19]: This class does not do anything on connection open. This comment references this block of code:

```C++
// Binance_Websockets.cpp

/* void StartDatabaseServer()
 *
 * This function sets the server's behavior on message callback over port 8081, and
 * starts the database's server.
 *
 * Callback from the frontend of MessageType::Open tells the Database_Handler that the frontend is ready,
 * prompting the Database_Handler to send all holdings data through to the frontend.
 * 
 *Callback from the frontend of MessageType::Message writes valid transactions to the database.csv file 
 * by contstruction a stringstream of the transaction data, and passing it to the Database_Handler's writeTransaction() method.
 *  
 * Valid transactions are JSON objects with fields: 
 * "type" - e.g. "buy" or "sell", 
 * "coin" - "BTC"
 * "price" - "73583"
 * "quantity" - "0.02156"
 * 
 * Important: StartServer() uses Server.listenAndStart(), which is a BLOCKING
 * call. If you are to later try and join a server thread, you must first call
 * Server.stop().
 */
void StartDatabaseServer() {
    DatabaseServer.setOnClientMessageCallback(
    [](shared_ptr<ix::ConnectionState> connectionState, WebSocket &webSocket, const MessagePtr &msg) {
        Database_Handler db;
        if (msg->type == MessageType::Open) {
            db.sendHoldingsData(webSocket);
            db.sendTransactionHistory(webSocket);
            db.sendHistoricalData(webSocket, "btcusdt");
            db.sendHistoricalData(webSocket, "ethusdt");
            db.sendHistoricalData(webSocket, "adausdt");
            db.sendHistoricalData(webSocket, "xrpusdt");
            db.sendHistoricalData(webSocket, "dotusdt");
            db.sendHistoricalData(webSocket, "uniusdt");
        }
        if (msg->type == MessageType::Message) {
                json incoming = json::parse(msg->str);
                string transactionType = incoming["type"].get<std::string>();
                if (transactionType == "buy" || transactionType == "sell") {
                    db.recordTransaction(&incoming);
                }
        }
    });
    DatabaseServer.listenAndStart();
}
```
In order for this class to do something on connection open, it would have to check and act on connection open. Rather, this class is called on connection open. Minor semantics BS comment on my end.

Could delete these three lines without losing meaning.

File: src/Database_Handler.h [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+            ofstream file("../userData/transactionHistory.csv", ios::app);
+            if (file.is_open()) {
+                file << type << "," 
+                     << coin << ","
+                     << toStr((*transactionData)["price"]) << ","
+                     << quantity << "," << time(0) << endl;
+            } else {
+                printf("Database_Handler: error writing to transactionHistory.csv\n"); // Excessive test logging
+            }
+        }
```
Comment (hwpoole) [inline src/Database_Handler.h:52]: I would either make everything a variable, then add that variable to the file, or skip variables entirely and do everything as you did for `["price"]`. Here, we have a mismatch of styles for doing this.

```C++
ofstream file("../userData/transactionHistory.csv", ios::app);
if (file.is_open()) {
    file << toStr((*transactionData)["type"]) << ","
         << toStr((*transactionData)["coin"]) << ","
         << toStr((*transactionData)["price"]) << ","
         << (*transactionData)["quantity"].get<double>() << ","
         << time(0) << endl;
} else {
// printf
}

Comment (hwpoole) [reply src/Database_Handler.h:52]: I see. You need the value of `type, coin` for later.

File: src/Database_Handler.h [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        {
+            ofstream file("../userData/holdings.csv", ios::trunc);
+            if (!file.is_open()) {
+                printf("Database_Handler: error writing to holdings.csv\n"); // Excessive test logging
+            } else {
+                for (auto& [mappedCoin, mappedQty] : holdings) {
+                    file << mappedCoin << "," << mappedQty << endl;
+                }
+            }
+        }
```
Comment (hwpoole) [inline src/Database_Handler.h:93]: This feels like it should be a private helper method.

Consider making this as follows:

```C++
class Database_Handler {
private:
    /* void updateHoldings(string type, string coin, double quantity) {...}
     *
     * This method writes the changes in quantity back to the holdings database.
     *
     * 1. Make a map<string, double> holdings.
     * 2. Open the file stream to holdings.csv
     *    2a. If that fails, print an error.
     * 3. Get each coin/quantity pair from holdings and add it to the map.
     * 4. Adjust quantity in the map based on buy/sell type and quantity.
     * 5. Write the updated holdings back to holdings.csv
     */
    void updateHoldings(string type, string coin, double quantity) {
        map<string, double> holdings;
        
        ifstream file("../userData/holdings.csv");
        if (!file.is_open()) {
            printf("Database_Handler: error reading holdings.csv\n"); // Excessive test logging
        } else {
            string line;
            while (getline(file, line)) {
                if (line.empty()) continue;
                stringstream ss(line);
                string mappedCoin, mappedQty;
                getline(ss, mappedCoin, ',');
                getline(ss, mappedQty, ',');
                holdings[mappedCoin] = atof(mappedQty.c_str());
            }
        }

        // Apply negative/positive behavior based on JSON field "type"
        if (type == "buy") {
            holdings[coin] += quantity;
        } else if (type == "sell") {
            holdings[coin] -= quantity;
            //* Boundary case to test coin removal, not permanent, as Bounds checking was supposed to be done in frontend 
            if (holdings[coin] <= 0) holdings.erase(coin); 
        }

        // Rewrite holdings.csv
        ofstream file("../userData/holdings.csv", ios::trunc);
        if (!file.is_open()) {
            printf("Database_Handler: error writing to holdings.csv\n"); // Excessive test logging
        } else {
            for (auto& [mappedCoin, mappedQty] : holdings) {
                file << mappedCoin << "," << mappedQty << endl;
            }
        }
    }

public:
    /* inline void recordTransaction()
     *
     * This method takes a json *transactionData and writes its contents to the transactionHistory.csv database. Then, it calls updateHoldings() to process the changes to the user's current holdings.
     */
    inline void recordTransaction(json* transactionData) {
        // Helper function to convert JSON values  to strings
        auto toStr = [](const json& value) -> std::string {
            if (value.is_string()) return value.get<std::string>();
            return value.dump(); // raw text for non-strings
        };

        string type = toStr((*transactionData)["type"]);
        string coin = toStr((*transactionData)["coin"]);
        double quantity = (*transactionData)["quantity"].get<double>();
    // Transactions:
    // Leverages ios append mode to write the incoming transactionto the end of the CSV database.
        ofstream file("../userData/transactionHistory.csv", ios::app);
        if (file.is_open()) {
            file << type << "," 
                 << coin << ","
                 << toStr((*transactionData)["price"]) << ","
                 << quantity << "," << time(0) << endl;
        } else {
            printf("Database_Handler: error writing to transactionHistory.csv\n"); // Excessive test logging
        }
        updateHoldings(type, coin, quantity);
    }
// Etc...
```

Recall the private methods can't be accessed from outside this class, so you're good to do whatever you want there without fear of exposing something you shouldn't.

File: src/Database_Handler.h [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        {
+            ifstream file("../userData/holdings.csv");
+            if (!file.is_open()) {
+                printf("Database_Handler: error reading holdings.csv\n"); // Excessive test logging
+                json empty;
+                empty["dataType"] = "holding";
+                empty["last"] = true;
+                webSocket.send(empty.dump());
+                return;
+            }
```
Comment (hwpoole) [inline src/Database_Handler.h:112]: Just out of curiosity, what's with the `{...}` that guard some code blocks? It is a stylistic thing? I haven't seen that before.

File: src/Database_Handler.h (hunk 8 above)
Comment (hwpoole) [reply src/Database_Handler.h:52]: You write strings here and it causes problems later for `sendTransactionHistory()`, where you have to get the string, convert it to a c style string, then call `atof` to get a double.

Why not just store a double where you need a double? `ofstream` accepts doubles:

```C++
ofstread file("../userData/transactionHistory.csv", ios::app);
if (file.is_open()) {
    file << toStr((*transactionData)["type"]) << ","
         << toStr((*transactionData)["coin"]) << ","
         << (*transactionData)["price"].get<double>() << ","
         << (*transactionData)["quantity"].get<double>() << ","
         << time(0) << endl;
} else {
//printf, etc...
```

File: src/Database_Handler.h [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+            getline(ss, price, ',');
+            getline(ss, qty, ',');
+            getline(ss, time, ',');
+
+            json entry;
+            entry["dataType"] = "transaction";
+            entry["type"] = type;
+            entry["coin"] = coin;
+            entry["price"] = atof(price.c_str());
+            entry["quantity"] = atof(qty.c_str());
```
Comment (hwpoole) [inline src/Database_Handler.h:179]: `entry["price"]` ... store it as a double rather than a string.
`entry["quantity"]` ... is already stored as a double.

1. Quantity is a double.
2. Quantity is converted to a C-style string.
3. Quantity is converted to a double.
4. Quantity is stored in `entry["quantity"] as a double.
5. ???
6. Profit.

```C++
entry["price"] = price;
entry["quantity"] = quantity;
// etc...
```

File: src/Database_Handler.h [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+            // marketCap and totalVolume are discarded, but we need to read them in to get to the next line
+            getline(ss, marketCap,   ','); 
+            getline(ss, totalVolume, ','); 
+
+            json entry;
+            entry["dataType"] = "historical";
+            entry["coin"] = coin;
+            entry["time"] = atol(timestamp.c_str());
+            entry["price"] = atof(price.c_str());
+            entry["last"] = false;
```
Comment (hwpoole) [inline src/Database_Handler.h:254]: This type of thing got typed a lot in this file. You could have made a JSON type for any and all JSONs that were being created frequently.

Check out these lines from `BinanceProcessor.h`

```C++
/* These two structs are used to define the shape of the JSON object we want to
 * create from Binance's response.
 *
 * We create struct KlineData to select, well, the KlineData, and struct
 * OutboundJSONStruct to hold the actual JSON we'd like to send out to the
 * frontend.
 *
 * Defining the shape here allows us to change the shape of the JSON object
 * without hunting down the specific lines of code in the client or
 * server. Hence the creation of this header file.
 *
 * Likely to be more useful later than they are now.
 */
struct KlineData {
  string Open;
  string Close;
  string High;
  string Low;
  bool KlineFinished;
};

struct OutboundJSONStruct {
  uint64_t EventTime;
  string Coin;
  KlineData Kline;
};

/* These lines tell nlohmann/json to define KlineData and OutboundJSONStruct as
 *  types.
 *
 *  Should be <type_you_want>, <member_1>, <member_2>, ..., <member_n>
 */
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(KlineData, Open, Close, High, Low);
NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(OutboundJSONStruct, EventTime, Coin, Kline);

// Now we have a full fat type that can even be a return type, as below.

/* toSimpleKline()
   *
   * This function takes a pointer to the response of a Binance WebSocket Kline
   * stream, simplifies it into the basic Kline facts, and then returns a
   * complete OutboundJSONStruct.
   */
  static OutboundJSONStruct toSimpleKline(const json &received) {
    OutboundJSONStruct outbound;
    outbound.EventTime = received["E"];
    outbound.Coin = received["s"];
    outbound.Kline.Open = received["k"]["o"];
    outbound.Kline.Close = received["k"]["c"];
    outbound.Kline.High = received["k"]["h"];
    outbound.Kline.Low = received["k"]["l"];
    outbound.Kline.KlineFinished = received["k"]["x"];
    return outbound;
  }
```

In your case, you could do this:

```C++
struct EntryType {
  string dataType; // or string dataType, coin;
  string coin;
  time_t time;
  double price;
  bool last;
};

NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(EntryType, dataType, coin, time, price, last);

class Database_Handler {
private:
// Stuff...
public:
// Things...

inline void sendHistoricalData(ix::WebSocket& webSocket, const string& coin) {
// Some code...

        while (getline(file, line)) {
            if (line.empty()) continue;
            stringstream ss(line);
            string timestamp, price, marketCap, totalVolume;

            getline(ss, timestamp,   ',');
            getline(ss, price,       ',');
            // marketCap and totalVolume are discarded, but we need to read them in to get to the next line
            getline(ss, marketCap,   ','); 
            getline(ss, totalVolume, ','); 

            EntryType entry;
            entry.dataType = "historical";
            entry.coin = coin;
            entry.time = timestamp;
            entry.price = price;
            entry.last = false;
            candles.push_back(entry);
      }

// Etc...
```

You could then do this for any other JSONs of known shape and size in your code. Now, you get all of the errors, warnings, and spellcheck you would get from a type. For instance, this is now _not_ allowed:

```C++
EntryType entry;
entry["Adding Something That Isn't Defined"] = crap; // Error!

// or

entry.SomeThing = OtherThing; // Error! No member 'SomeThing' in type EntryType.
```
It makes no functional difference but makes your life easier. If you wanted to add or delete a member of the `entry` JSON, you now only need to add it or delete it from the type `EntryType`, and it will be reflected everywhere that `EntryType` is used.

File: src/Binance_Websockets.cpp [hunk 13]
```diff
@@ -56,8 +63,54 @@ void StartServer() {
...
+        Database_Handler db;
+        if (msg->type == MessageType::Open) {
+            db.sendHoldingsData(webSocket);
+            db.sendTransactionHistory(webSocket);
+            db.sendHistoricalData(webSocket, "btcusdt");
+            db.sendHistoricalData(webSocket, "ethusdt");
+            db.sendHistoricalData(webSocket, "adausdt");
+            db.sendHistoricalData(webSocket, "xrpusdt");
+            db.sendHistoricalData(webSocket, "dotusdt");
+            db.sendHistoricalData(webSocket, "uniusdt");
```
Comment (hwpoole) [inline src/Binance_Websockets.cpp:103]: Could move `vector<string> coins` to be a global declaration and then use an enhanced for loop here:

```C++
for (string coin : coins) {
    db.sendHistoricalData(webSocket, coin);
}
```

Or even pass in the vector to StartDataBaseServer somehow. Then, there is no need to modify this if the number of coins change.

**Review by hwpoole — APPROVED** [review]

**Comment by hwpoole** [comment]
It's neither here nor there since this version works, but you very well could have just stored JSONs for Holdings and TransactionHistory. Did you consider doing that? Maybe using strings into a csv was easier?

File: src/Database_Handler.h [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            }
+        }
+
+        if (userHoldings.empty()) {
+            json empty;
+            empty["dataType"] = "holding";
+            empty["last"] = true;
+            webSocket.send(empty.dump());
+            return;
+        }
```
Comment (hwpoole) [inline src/Database_Handler.h:130]: You can use the `ifstream` method `peek()` to check if a file is empty from the get-go.

```C++
if (file.peek() == ifstream::traits_type::eof()) {
   // File is empty.
   // The first char in the file is the "end of file" type.
}
```

So, maybe this is:

```C++
if (!file.is_open()){
   // Things.
} else if (file.peek() == ifstream::traits_type::eof()) {
   // File is empty
} else {
   // Do work on file.
}
```
